packages/core-ethereum/crates/core-ethereum-rpc/src/indexer.rs

```diff
@@ -0,0 +1,360 @@
+
+                        // This will be only approximate, since we have no way of tracking how does the Stream internally make additional queries
+                        #[cfg(feature = "prometheus")]
+                        METRIC_COUNT_RPC_CALLS.increment();
+
+                        let mut current_block_log = BlockWithLogs { block_id: from_block, ..Default::default()};
+                        while let Ok(Some(log)) = retrieved_logs.try_next().await {
+                            let log = Log::from(log);
+
+                            // This assumes the logs are arriving ordered by blocks
+                            if current_block_log.block_id != log.block_number {
+                                debug!("completed {current_block_log}");
+                                yield current_block_log;
+
+                                current_block_log = BlockWithLogs::default();
+                                current_block_log.block_id = log.block_number;
+                            }
+
+                            debug!("retrieved {log}");
+                            current_block_log.logs.push(log);
+                        }
+
+                        debug!("retrieved complete {current_block_log}");
+                        yield current_block_log;
+                    }
+                    Err(e) => error!("failed to obtain current block number from chain: {e}")
+                }
+
+                sleep(self.cfg.expected_block_time).await;
+            }
+        }))
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use std::time::Duration;
+
+    use ethers::contract::EthEvent;
+    use ethers_providers::Middleware;
+    use futures::StreamExt;
+
+    use bindings::hopr_channels::*;
+    use bindings::hopr_token::{ApprovalFilter, HoprToken, TransferFilter};
+    use core_crypto::keypairs::{ChainKeypair, Keypair};
+    use core_ethereum_types::{create_anvil, create_rpc_client_to_anvil, ContractAddresses, ContractInstances};
+    use utils_log::debug;
+    use utils_types::primitives::Address;
+
+    use crate::client::tests::ReqwestRequestor;
+    use crate::client::{JsonRpcProviderClient, SimpleJsonRpcRetryPolicy};
+    use crate::rpc::tests::mint_tokens;
+    use crate::rpc::{RpcOperations, RpcOperationsConfig};
+    use crate::{BlockWithLogs, HoprIndexerRpcOperations, LogFilter};
+
+    async fn fund_channel<M: Middleware + 'static>(
+        counterparty: Address,
+        hopr_token: HoprToken<M>,
+        hopr_channels: HoprChannels<M>,
+    ) {
+        hopr_token
+            .approve(hopr_channels.address(), 1u128.into())
+            .send()
+            .await
+            .unwrap()
+            .await
+            .unwrap();
+
+        hopr_channels
+            .fund_channel(counterparty.into(), 1u128)
+            .send()
+            .await
+            .unwrap()
+            .await
+            .unwrap();
+    }
+
+    #[tokio::test]
```

Comment:
Why are tokio tests used, when the rest of the app and tests uses `async_std`?

Reply:
Required by ethers